packages/block-library/src/navigation/view.js

```diff
@@ -185,5 +190,26 @@ const { state, actions } = store( 'core/navigation', {
 				focusableElements?.[ 0 ]?.focus();
 			}
 		},
+		initNav() {
+			const context = getContext();
+			const mediaQuery = window.matchMedia(
+				`(max-width: ${ NAVIGATION_MOBILE_COLLAPSE })`
+			);
+
+			// Run once to set the initial state.
+			context.isCollapsed = mediaQuery.matches;
+
+			function handleCollapse( event ) {
+				context.isCollapsed = event.matches;
+			}
+
+			// Run on resize to update the state.
+			mediaQuery.addEventListener( 'change', handleCollapse );
+
+			// Remove the listener when the component is unmounted.
+			return () => {
+				mediaQuery.removeEventListener( 'change', handleCollapse );
+			};
+		},
```

Comment:
This seems to work. However the directives stuff might be good to get a review from @WordPress/interactivity-api folks.

Reply:
Sorry, I didn't comment here, but as @scruffian [said](https://github.com/WordPress/gutenberg/pull/57520#issuecomment-1877833183), this approach looks good to me 🙂